コメント表示機能を追加

mterada1228 opened the pull request:
- 変更の目的
  - コメント一覧を表示する
- 変更の詳細
  - `CommentsController#index` 時に `HorseRace` に紐づくコメントを取得して表示
  - クエリパラメータより取得する `comment_type` に該当するコメントのみが抽出される

mterada1228 committed:
コメント表示機能を追加

mterada1228 commented on app/views/comments/index.html.erb:
```diff
@@ -20,19 +11,28 @@
         <div class="form-group comment-header-area">
           <%= f.text_field :user_name, class: 'form-control', placeholder: 'ニックネームを入力' %>
         </div>
-        <div class="form-group comment-header-area">
```
処理が複雑になるため、`comment_type` はクエリパラメータのみ取得するようにして、ページからの選択はできないように修正しました。

ershn replied:
OK

ershn commented on app/controllers/comments_controller.rb:
```diff
@@ -1,6 +1,8 @@
 class CommentsController < ApplicationController
   def index
     @comment_type = params[:comment_type]
```
こちらのパラメーターのバリデーションをしないと、存在しないものを指定した場合に挙動がおかしくなるかと思います。

mterada1228 replied:
`validate_comment_type` を使用してバリデーションを行なっています。

ershn replied:
OK

ershn commented on app/controllers/comments_controller.rb:
```diff
@@ -1,6 +1,8 @@
     @comment_type = params[:comment_type]
+    @comments = HorseRace.find(params[:horse_race_id])
+                         .comments.where(comment_type: params[:comment_type].to_sym)
     @comment = Comment.new
```
`comments`と`comment`だと、用途の違いがぱっとわからないので、もっと具体的な名前にした方がいいと思いました。

`existing_comments`と`new_comment`がいいかと。

mterada1228 replied:
修正しました。

ershn replied:
OK

ershn commented on app/views/comments/index.html.erb:
```diff
@@ -20,19 +11,28 @@
+      <% end %>
+      コメント一覧
+    </h2>
+    <div class="comtainer">
```
`container`ですね。

mterada1228 replied:
修正しました。

ershn replied:
OK

ershn commented on app/views/comments/_comment.html.erb:
```diff
@@ -0,0 +1,10 @@
+    </div>
+    <div class="card-body">
+      <p><%= comment.description %></p>
+    </div>
```
セキュリティの問題ですが、こちらの値を DB に保存する際は事前に html タグの削除が必要です。

mterada1228 replied:
`create` アクションの前に、ユーザーがフリー入力できる項目は、`sanitize` するように修正しました。

ershn replied:
OK

mterada1228 committed:
指摘対応
  - DB に保存前に sanitize する
  - comment, comments は new_comment, exixting_comments に名称を変更

mterada1228 commented on config/application.rb:
```diff
@@ -33,5 +33,9 @@ class Application < Rails::Application
+
+    # ユーザー入力を許可する HTML タグを設定
+    config.action_view.sanitized_allowed_tags = []
+    config.action_view.sanitized_allowed_attributes = []
```
デフォルトの `sanitize` では、多くのHTMLを許可してしまうため、オーバーライドしています。

cf.)デフォルトで許可するタグ、アトリビュート
https://github.com/rails/rails-html-sanitizer/blob/db9ccec4d4f6756488c59c89afd3a414777fdea5/lib/rails/html/sanitizer.rb#L108-L111

ershn replied:
OK

ershn commented on app/controllers/comments_controller.rb:
```diff
@@ -21,4 +25,9 @@ def create
+  def sanitize_params
+    params[:comment][:user_name] = helpers.sanitize(params[:comment][:user_name])
+    params[:comment][:description] = helpers.sanitize(params[:comment][:description])
+  end
```
sanitize がデフォルトで様々な html タグを許容するのであれば、もっと適切なメソッドを使った方がいいと思います。

ここは strip_tags がいいかと。

mterada1228 replied:
修正しました。

ershn replied:
OK

ershn commented on app/controllers/comments_controller.rb:
```diff
@@ -21,4 +25,9 @@ def create
+  def sanitize_params
+    params[:comment][:user_name] = helpers.sanitize(params[:comment][:user_name])
+    params[:comment][:description] = helpers.sanitize(params[:comment][:description])
+  end
```
sanitize_params で現在 comment_params ではなく、params が使われてるので、パラメーターが渡されなかった場合は sanitize_params で想定外なエラーが発生してしまうと思います。

ここは comment_params と sanitize_params を一つのメソッドに統合した方がシンプルで使いやすいと思います。

mterada1228 replied:
修正しました。

ershn replied:
OK

mterada1228 committed:
Merge branch 'develop' into features/show_comments

mterada1228 committed:
指摘対応
  - 存在しない comment_type を指定した場合のバリデーションを追加する

mterada1228 committed:
指摘対応
  - sanitize ではなく、strip_tags を使用する
  - sanitaze_params ではなく、strong pamameters 内で html タグの除去を行う

mterada1228 committed:
行末に空行を挿入

ershn commented on app/controllers/comments_controller.rb:
```diff
@@ -19,6 +22,15 @@ def create
+    return if (Comment.comment_types.keys).include?(params[:comment_type])
+
+    render template: 'errors/error_400', status: 400
   end
```
パラメーターのバリデーションのために、ここはライブラリーを生かした方がいろいろ楽で安全だと思いました。

https://github.com/nicolasblanco/rails_param

mterada1228 replied:
こちら対応いたしました。

ershn replied:
OK

mterada1228 committed:
指摘対応
  - rails_param gem の導入

ershn commented on app/controllers/comments_controller.rb:
```diff
@@ -19,6 +28,7 @@ def create
   def comment_params
-    params.require(:comment).permit(:description, :user_name, :comment_type)
+    params.require(:comment).permit(:description, :user_name)
+          .merge({ comment_type: params[:comment_type] })
```
rails_param を使うなら、strong parameters は使う必要がないと思います。

rails_param は`params`をそのまま更新するので、`params`を直接使って問題ないです。

mterada1228 replied:
Rails 4 以降で、`strong_parameters` gem が default で有効になったことで、明示的に `permit` するカラムを指定しない場合、`create`, `update` のアクションが行えなくなったようです。なので、そのまま `strong_parameters` を残す形としたいのですが、問題ないでしょうか。

参考）https://stackoverflow.com/questions/17450185/forbidden-attributes-error-in-rails-4-when-encountering-a-situation-where-one-wo

ershn replied:
OK

ershn commented on app/controllers/comments_controller.rb:
```diff
@@ -1,15 +1,24 @@
+    param! :comment, Hash do |c|
+      c.param! :description, String, required: true, transform: ->(v) { helpers.strip_tags(v) }
+      c.param! :user_name,   String, required: true, transform: ->(v) { helpers.strip_tags(v) }
+    end
```
- `comment_type`と`comment`は必須だと思いますので、`required: true`にした方が適切かと
- パラメーターのバリデーションは個別なメソッドで実装した方がメソッドの間で単一責任の原則が守られます

mterada1228 replied:
こちら修正いたしました。

ershn replied:
OK

ershn commented on app/controllers/comments_controller.rb:
```diff
@@ -1,15 +1,24 @@
   def index
-    @comment_type = params[:comment_type]
-    @comment = Comment.new
+    param! :comment_type, String, in: Comment.comment_types.keys
```
`to_sym`は`transform`でも指定できると思います。

mterada1228 replied:
こちら修正いたしました。

ershn replied:
OK

mterada1228 committed:
comment_type もフォームとして受け取るように変更

mterada1228 committed:
params のバリデーションはメソッドを別に切り出す

mterada1228 committed:
rails_params のバリデーションを修正

mterada1228 committed:
CommentsController#create はクエリパラメータを受け取る必要がなくなったので修正する

ershn commented on app/controllers/comments_controller.rb:
```diff
@@ -21,4 +25,16 @@ def create
   end
+
+  def validate_query
+    param! :comment_type, String, in: Comment.comment_types.keys, trandform: ->(v) { v.to_sym }
```
`transform`ですね。

もしかすると、`to_sym`は必要ないです。

mterada1228 replied:
ご指摘の通り、シンボルに変換する必要がなかったので削除しました。

ershn replied:
OK

ershn commented on app/controllers/comments_controller.rb:
```diff
@@ -1,15 +1,19 @@
     horse_race = HorseRace.find(params[:horse_race_id])
-    @comment = horse_race.comments.new(comment_params)
-    if @comment.save
+    @new_comment = horse_race.comments.new(comment_params)
```
`strong_parameters`を使う必要がないように、ここは`params`を直接渡すのではなくて、個別で新しいハッシュを作れば大丈夫だと思います。

```ruby
params[:comment].slice(:description, :user_name, :comment_type)
```

mterada1228 replied:
ご指摘いただいた方法ですと、結局は `permit` する必要があります。

``` ruby
params[:comment].slice(:description, :user_name, :comment_type)
=> <ActionController::Parameters {"description"=>"aaa", "user_name"=>"aaa", "comment_type"=>"positive"} permitted: false>
```

`slice` で取得されるのは、`ActionControlle::Parameters` クラスで、メンバー変数の `permitted` が `false` のためです。これを `ActiveRecord::Relation#new` 等に渡すと、`ActiveModel::ForbiddenAttributesError` が raise されます。

なので、`comment_params` で全く新しいハッシュを生成して、新しいハッシュを initial data として `comment` インスタンスを `new` する方法としました。

ershn replied:
OK

ershn commented on app/controllers/comments_controller.rb:
```diff
@@ -21,4 +25,16 @@ def create
+  def validate_comment_params
+    param! :comment, Hash, required: true do |c|
+      c.param! :description,  String, required: true, transform: ->(v) { helpers.strip_tags(v) }
+      c.param! :user_name,    String, required: true, transform: ->(v) { helpers.strip_tags(v) }
```
ここは`required`が必要ないと思います。

mterada1228 replied:
こちら削除しました。

ershn replied:
OK

ershn commented on app/controllers/comments_controller.rb:
```diff
@@ -1,15 +1,19 @@
 class CommentsController < ApplicationController
+  before_action :validate_query, only: [:index]
+  before_action :validate_comment_params, only: [:create]
```
命名ですが、アクションに倣ってメソッドを名付けてもいいと思いました。

```
validate_index_params
validate_create_params
```

mterada1228 replied:
こちら修正いたしました。

ershn replied:
OK

mterada1228 committed:
指摘対応
